File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int l=0,r=0,maxF=0,maxL=0;
        int a[]=new int[26];
        while(r<s.length()){
            a[s.charAt(r)-'A']++;
            maxF=Math.max(maxF,a[s.charAt(r)-'A']);
            while((r-l+1)-maxF>k){
                a[s.charAt(l)-'A']--;
                l++;
            }
            maxL=Math.max(maxL,r-l+1);
            r++;
        }
        return maxL;
    }
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        dic={}
        l=r=maxl=maxf=0
        while r<len(s):
            dic[s[r]]=dic.get(s[r],0)+1
            maxf=max(maxf,dic[s[r]])
            changes=(r-l+1)-maxf
            while changes>k:
                dic[s[l]] -= 1
                if dic[s[l]] == 0:
                    del dic[s[l]]
                l += 1
                maxf=max(maxf,dic[s[r]])
                changes=(r-l+1)-maxf
            maxl=max(maxl,r-l+1)
            r+=1

        return maxl

                
